File: src/history_file.c
/*
 * history_file.c - Flat-file based input history (one entry per line)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/stat.h>
#include <limits.h>
#include <unistd.h>

#include "history_file.h"
#include "logger.h"

struct HistoryFile {
    char *path;
    FILE *fp; // append handle
};

static int mkdir_recursive(const char *path) {
    char tmp[PATH_MAX];
    char *p = NULL;
    size_t len;

    snprintf(tmp, sizeof(tmp), "%s", path);
    len = strlen(tmp);
    if (len == 0) return -1;
    if (tmp[len - 1] == '/') tmp[len - 1] = '\0';
    for (p = tmp + 1; *p; p++) {
        if (*p == '/') {
            *p = '\0';
            if (mkdir(tmp, 0755) != 0 && errno != EEXIST) return -1;
            *p = '/';
        }
    }
    if (mkdir(tmp, 0755) != 0 && errno != EEXIST) return -1;
    return 0;
}

char* history_file_default_path(void) {
    char *path = NULL;
    const char *env = getenv("CLAUDE_C_HISTORY_FILE_PATH");
    if (env && *env) return strdup(env);

    struct stat st;
    if (stat("./.claude-c", &st) == 0 && S_ISDIR(st.st_mode)) {
        return strdup("./.claude-c/input_history.txt");
    }
    if (mkdir("./.claude-c", 0755) == 0 || errno == EEXIST) {
        return strdup("./.claude-c/input_history.txt");
    }

    const char *xdg = getenv("XDG_DATA_HOME");
    if (xdg && *xdg) {
        path = malloc(PATH_MAX);
        if (path) {
            snprintf(path, PATH_MAX, "%s/claude-c/input_history.txt", xdg);
            return path;
        }
    }

    const char *home = getenv("HOME");
    if (home && *home) {
        path = malloc(PATH_MAX);
        if (path) {
            snprintf(path, PATH_MAX, "%s/.local/share/claude-c/input_history.txt", home);
            return path;
        }
    }
    return strdup("./input_history.txt");
}

HistoryFile* history_file_open(const char *path) {
    HistoryFile *hf = calloc(1, sizeof(HistoryFile));
    if (!hf) return NULL;

    if (path && *path) hf->path = strdup(path);
    else hf->path = history_file_default_path();
    if (!hf->path) { free(hf); return NULL; }

    char dir[PATH_MAX];
    snprintf(dir, sizeof(dir), "%s", hf->path);
    char *slash = strrchr(dir, '/');
    if (slash) { *slash = '\0'; mkdir_recursive(dir); }

    hf->fp = fopen(hf->path, "a");
    if (!hf->fp) {
        LOG_WARN("[HIST] Failed to open history file for append: %s", hf->path);
        free(hf->path);
        free(hf);
        return NULL;
    }
    // Use line buffering; no fsync to keep things fast
    setvbuf(hf->fp, NULL, _IOLBF, 0);
    return hf;
}

void history_file_close(HistoryFile *hf) {
    if (!hf) return;
    if (hf->fp) fclose(hf->fp);
    free(hf->path);
    free(hf);
}

// Escape newlines in text for storage
#ifdef TEST_BUILD
char* escape_newlines(const char *text) {
#else
static char* escape_newlines(const char *text) {
#endif
    if (!text) return NULL;

    // Count how many newlines we need to escape
    size_t newline_count = 0;
    for (const char *p = text; *p; p++) {
        if (*p == '\n') newline_count++;
    }

    // If no newlines, just return a copy
    if (newline_count == 0) {
        return strdup(text);
    }

    // Allocate buffer for escaped text (each \n becomes \\n)
    size_t escaped_len = strlen(text) + newline_count + 1;
    char *escaped = malloc(escaped_len);
    if (!escaped) return NULL;

    // Perform escaping
    char *dest = escaped;
    for (const char *src = text; *src; src++) {
        if (*src == '\n') {
            *dest++ = '\\';
            *dest++ = 'n';
        } else {
            *dest++ = *src;
        }
    }
    *dest = '\0';

    return escaped;
}

// Unescape newlines when loading from storage
#ifdef TEST_BUILD
char* unescape_newlines(const char *escaped_text) {
#else
static char* unescape_newlines(const char *escaped_text) {
#endif
    if (!escaped_text) return NULL;

    // Count how many escape sequences we have
    size_t escape_count = 0;
    for (const char *p = escaped_text; *p; p++) {
        if (*p == '\\' && *(p + 1) == 'n') {
            escape_count++;
            p++; // Skip the 'n'
        }
    }

    // If no escapes, just return a copy
    if (escape_count == 0) {
        return strdup(escaped_text);
    }

    // Allocate buffer for unescaped text (each \\n becomes \n)
    size_t unescaped_len = strlen(escaped_text) - escape_count + 1;
    char *unescaped = malloc(unescaped_len);
    if (!unescaped) return NULL;

    // Perform unescaping
    char *dest = unescaped;
    for (const char *src = escaped_text; *src; src++) {
        if (*src == '\\' && *(src + 1) == 'n') {
            *dest++ = '\n';
            src++; // Skip the 'n'
        } else {
            *dest++ = *src;
        }
    }
    *dest = '\0';

    return unescaped;
}

int history_file_append(HistoryFile *hf, const char *text) {
    if (!hf || !hf->fp || !text || text[0] == '\0') return 0;

    // Escape newlines before storing
    char *escaped_text = escape_newlines(text);
    if (!escaped_text) {
        LOG_WARN("[HIST] Failed to escape text for history");
        return -1;
    }

    if (fprintf(hf->fp, "%s\n", escaped_text) < 0) {
        LOG_WARN("[HIST] Failed to append to history file: %s", hf->path);
        free(escaped_text);
        return -1;
    }
    fflush(hf->fp); // allow OS buffering; no fsync
    free(escaped_text);
    return 0;
}

char** history_file_load_recent(HistoryFile *hf, int limit, int *out_count) {
    if (out_count) *out_count = 0;
    if (!hf || !hf->path || limit <= 0) return NULL;

    FILE *fr = fopen(hf->path, "r");
    if (!fr) return NULL;

    // Read entire file into memory (simple; history file typically small)
    if (fseek(fr, 0, SEEK_END) != 0) { fclose(fr); return NULL; }
    long size = ftell(fr);
    if (size < 0) { fclose(fr); return NULL; }
    if (fseek(fr, 0, SEEK_SET) != 0) { fclose(fr); return NULL; }

    char *buf = malloc((size_t)size + 1);
    if (!buf) { fclose(fr); return NULL; }
    size_t nread = fread(buf, 1, (size_t)size, fr);
    fclose(fr);
    buf[nread] = '\0';

    // Count lines
    int total_lines = 0;
    for (size_t i = 0; i < nread; i++) if (buf[i] == '\n') total_lines++;
    if (nread > 0 && buf[nread - 1] != '\n') total_lines++; // last line without newline

    if (total_lines == 0) { free(buf); return NULL; }

    // Determine start line to keep only last 'limit'
    int start_line = total_lines > limit ? total_lines - limit : 0;

    // Collect lines
    char **tmp = calloc((size_t)(total_lines - start_line), sizeof(char*));
    if (!tmp) { free(buf); return NULL; }

    int line_idx = 0, out_idx = 0;
    char *line_start = buf;
    for (size_t i = 0; i <= nread; i++) {
        if (i == nread || buf[i] == '\n') {
            size_t len = (size_t)(&buf[i] - line_start);
            // At end of buffer (i == nread): only process if there's content after the last \n
            // This prevents processing a phantom empty line when file ends with \n
            int should_process = (i < nread) || (len > 0);

            if (should_process && line_idx >= start_line) {
                // Trim carriage return
                if (len > 0 && line_start[len - 1] == '\r') len--;
                if (len > 0) {
                    // Create null-terminated copy of the line
                    char *line_copy = malloc(len + 1);
                    if (line_copy) {
                        memcpy(line_copy, line_start, len);
                        line_copy[len] = '\0';

                        // Unescape newlines
                        char *unescaped = unescape_newlines(line_copy);
                        free(line_copy);

                        if (unescaped) {
                            tmp[out_idx++] = unescaped;
                        }
                    }
                }
            }
            if (should_process) {
                line_idx++;
            }
            line_start = (i < nread) ? &buf[i + 1] : &buf[i];
        }
    }

    free(buf);
    if (out_count) *out_count = out_idx;
    return (out_idx > 0) ? tmp : (free(tmp), NULL);
}

File: zig/history_file.zig
//! history_file.zig — Flat-file input history (one entry per line)
//!
//! Zig port of src/history_file.c.
//!
//! Each entry is stored on a single physical line with `\n` characters
//! escaped as the two-character sequence `\n` (backslash + 'n').
//! This lets multi-line editor inputs survive a round-trip through the file.

const std = @import("std");
const data_dir = @import("data_dir.zig");

/// Escape `\n` → `\n` (two chars) for safe single-line storage.
/// Returns a newly-allocated string; caller must free.
pub fn escapeNewlines(allocator: std.mem.Allocator, text: []const u8) ![]u8 {
    // Count newlines to pre-size the buffer.
    var count: usize = 0;
    for (text) |ch| if (ch == '\n') { count += 1; };

    if (count == 0) return allocator.dupe(u8, text);

    var buf = try allocator.alloc(u8, text.len + count);
    var out: usize = 0;
    for (text) |ch| {
        if (ch == '\n') {
            buf[out] = '\\';
            out += 1;
            buf[out] = 'n';
        } else {
            buf[out] = ch;
        }
        out += 1;
    }
    return buf[0..out];
}

/// Unescape `\n` (two chars) → `\n` when loading from storage.
/// Returns a newly-allocated string; caller must free.
pub fn unescapeNewlines(allocator: std.mem.Allocator, text: []const u8) ![]u8 {
    // Count `\n` sequences.
    var count: usize = 0;
    var i: usize = 0;
    while (i < text.len) : (i += 1) {
        if (text[i] == '\\' and i + 1 < text.len and text[i + 1] == 'n') {
            count += 1;
            i += 1; // skip 'n'
        }
    }

    if (count == 0) return allocator.dupe(u8, text);

    var buf = try allocator.alloc(u8, text.len - count);
    var out: usize = 0;
    i = 0;
    while (i < text.len) : (i += 1) {
        if (text[i] == '\\' and i + 1 < text.len and text[i + 1] == 'n') {
            buf[out] = '\n';
            out += 1;
            i += 1;
        } else {
            buf[out] = text[i];
            out += 1;
        }
    }
    return buf[0..out];
}

/// Resolve the default history file path.
///
/// Priority: `$KLAWED_HISTORY_FILE_PATH` → data_dir/input_history.txt
/// → `$XDG_DATA_HOME/klawed/input_history.txt`
/// → `$HOME/.local/share/klawed/input_history.txt`
/// → `./input_history.txt`
///
/// Returns a caller-owned string.
pub fn defaultPath(allocator: std.mem.Allocator) ![]const u8 {
    // 1. Explicit env override.
    if (std.process.getEnvVarOwned(allocator, "KLAWED_HISTORY_FILE_PATH")) |p| {
        if (p.len > 0) return p;
        allocator.free(p);
    } else |_| {}

    // 2. Project-local data directory.
    if (data_dir.buildPath(allocator, "input_history.txt")) |p| {
        return p;
    } else |_| {}

    // 3. XDG_DATA_HOME.
    if (std.process.getEnvVarOwned(allocator, "XDG_DATA_HOME")) |xdg| {
        defer allocator.free(xdg);
        if (xdg.len > 0) {
            return std.fs.path.join(allocator, &.{ xdg, "klawed", "input_history.txt" });
        }
    } else |_| {}

    // 4. HOME fallback.
    if (std.process.getEnvVarOwned(allocator, "HOME")) |home| {
        defer allocator.free(home);
        if (home.len > 0) {
            return std.fs.path.join(
                allocator,
                &.{ home, ".local", "share", "klawed", "input_history.txt" },
            );
        }
    } else |_| {}

    // 5. CWD fallback.
    return allocator.dupe(u8, "./input_history.txt");
}

/// An open history file that supports appending and loading recent entries.
pub const HistoryFile = struct {
    path: []const u8, // owned
    file: std.fs.File,
    allocator: std.mem.Allocator,

    /// Open (or create) the history file at `path`.  If `path` is empty, the
    /// default path is used.  The file is opened in append mode.
    pub fn open(allocator: std.mem.Allocator, path: ?[]const u8) !HistoryFile {
        const resolved_path = if (path != null and path.?.len > 0)
            try allocator.dupe(u8, path.?)
        else
            try defaultPath(allocator);
        errdefer allocator.free(resolved_path);

        // Ensure parent directory exists.
        if (std.fs.path.dirname(resolved_path)) |dir| {
            std.fs.cwd().makePath(dir) catch {};
        }

        const file = try std.fs.cwd().createFile(resolved_path, .{
            .read = false,
            .truncate = false,
        });
        // Seek to end for append semantics.
        try file.seekFromEnd(0);

        return HistoryFile{
            .path = resolved_path,
            .file = file,
            .allocator = allocator,
        };
    }

    /// Close the history file and free resources.
    pub fn close(self: *HistoryFile) void {
        self.file.close();
        self.allocator.free(self.path);
        self.* = undefined;
    }

    /// Append a single entry.  Empty strings are silently ignored.
    pub fn append(self: *HistoryFile, text: []const u8) !void {
        if (text.len == 0) return;

        const escaped = try escapeNewlines(self.allocator, text);
        defer self.allocator.free(escaped);

        var bw = std.io.bufferedWriter(self.file.writer());
        try bw.writer().writeAll(escaped);
        try bw.writer().writeByte('\n');
        try bw.flush();
    }

    /// Load the `limit` most-recent entries.
    ///
    /// Returns a slice of caller-owned strings; use `freeLoaded` to release.
    /// Returns an empty slice when the file is empty or unreadable.
    pub fn loadRecent(
        self: *const HistoryFile,
        limit: usize,
    ) ![][]const u8 {
        if (limit == 0) return &[_][]const u8{};

        // Read the entire file.
        const file = std.fs.cwd().openFile(self.path, .{}) catch
            return &[_][]const u8{};
        defer file.close();

        const content = try file.readToEndAlloc(self.allocator, 64 * 1024 * 1024);
        defer self.allocator.free(content);

        // Split into lines.
        var lines = std.ArrayList([]const u8).init(self.allocator);
        defer {
            for (lines.items) |l| self.allocator.free(l);
            lines.deinit();
        }

        var it = std.mem.splitScalar(u8, content, '\n');
        while (it.next()) |raw_line| {
            // Trim trailing \r.
            var line = raw_line;
            if (line.len > 0 and line[line.len - 1] == '\r') {
                line = line[0 .. line.len - 1];
            }
            if (line.len == 0) continue;

            const unescaped = try unescapeNewlines(self.allocator, line);
            try lines.append(unescaped);
        }

        // Return only the last `limit` entries.
        const total = lines.items.len;
        const start = if (total > limit) total - limit else 0;
        const slice = lines.items[start..];

        var result = try self.allocator.alloc([]const u8, slice.len);
        for (slice, 0..) |l, i| {
            result[i] = l;
            // Ownership transferred — prevent the defer above from freeing.
            lines.items[start + i] = &[_]u8{};
        }
        return result;
    }

    /// Free the slice returned by `loadRecent`.
    pub fn freeLoaded(self: *const HistoryFile, entries: [][]const u8) void {
        for (entries) |e| self.allocator.free(e);
        self.allocator.free(entries);
    }
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "escape and unescape round-trip" {
    const allocator = std.testing.allocator;

    const original = "line one\nline two\nline three";
    const escaped = try escapeNewlines(allocator, original);
    defer allocator.free(escaped);

    // Escaped should not contain literal newlines.
    try std.testing.expect(std.mem.indexOf(u8, escaped, "\n") == null);

    const restored = try unescapeNewlines(allocator, escaped);
    defer allocator.free(restored);
    try std.testing.expectEqualStrings(original, restored);
}

test "escape no-op for strings without newlines" {
    const allocator = std.testing.allocator;
    const s = "hello world";
    const escaped = try escapeNewlines(allocator, s);
    defer allocator.free(escaped);
    try std.testing.expectEqualStrings(s, escaped);
}

test "unescape no-op for strings without escape sequences" {
    const allocator = std.testing.allocator;
    const s = "hello world";
    const result = try unescapeNewlines(allocator, s);
    defer allocator.free(result);
    try std.testing.expectEqualStrings(s, result);
}

test "HistoryFile append and loadRecent" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const allocator = std.testing.allocator;

    // Build an absolute path inside the tmp dir.
    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const hist_path = try tmp.dir.realpath(".", &path_buf);
    const full_path = try std.fs.path.join(allocator, &.{ hist_path, "history.txt" });
    defer allocator.free(full_path);

    var hf = try HistoryFile.open(allocator, full_path);
    defer hf.close();

    try hf.append("first entry");
    try hf.append("second entry");
    try hf.append("multi\nline\nentry");

    const entries = try hf.loadRecent(10);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 3), entries.len);
    try std.testing.expectEqualStrings("first entry", entries[0]);
    try std.testing.expectEqualStrings("second entry", entries[1]);
    try std.testing.expectEqualStrings("multi\nline\nentry", entries[2]);
}

test "HistoryFile loadRecent respects limit" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();

    const allocator = std.testing.allocator;

    var path_buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
    const dir_path = try tmp.dir.realpath(".", &path_buf);
    const full_path = try std.fs.path.join(allocator, &.{ dir_path, "history2.txt" });
    defer allocator.free(full_path);

    var hf = try HistoryFile.open(allocator, full_path);
    defer hf.close();

    var i: usize = 0;
    while (i < 5) : (i += 1) {
        const line = try std.fmt.allocPrint(allocator, "entry {d}", .{i});
        defer allocator.free(line);
        try hf.append(line);
    }

    const entries = try hf.loadRecent(3);
    defer hf.freeLoaded(entries);

    try std.testing.expectEqual(@as(usize, 3), entries.len);
    try std.testing.expectEqualStrings("entry 2", entries[0]);
    try std.testing.expectEqualStrings("entry 3", entries[1]);
    try std.testing.expectEqualStrings("entry 4", entries[2]);
}
